### PR TITLE
Add bindings for BLS from Chia network 

### PR DESCRIPTION
**Pull request by mhchia**
### What was wrong?
https://github.com/ethereum/py_ecc/issues/71

### How was it fixed?
Moved from https://github.com/ethereum/py_ecc/pull/72.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by mhchia** [comment]
CI failed in
- [docker-image-build-test](https://circleci.com/gh/mhchia/trinity/2976), due to `bls-bindings` is added in `dev` in `setup.py`.
- [py36-rpc-blockchain](https://circleci.com/gh/mhchia/trinity/3000), due to unknown reason.

**Comment by mhchia** [comment]
Another unknown CI failure: https://circleci.com/gh/mhchia/trinity/3002?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by carver** [comment]
Sorry, flaky CI is terrible to work with :(

Working on it:



> * [py36-rpc-blockchain](https://circleci.com/gh/mhchia/trinity/3000), due to unknown reason.

Yeah, there was a bad eth-keys release. This PR will need to be merged and released and then the test will be resolved: https://github.com/ethereum/eth-keys/pull/62

> Another unknown CI failure: https://circleci.com/gh/mhchia/trinity/3002?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Ugh, I haven't seen this one trigger before, but it isn't surprising, since it's sensitive to the VMs speed. We just need to increase the tolerance, I guess. Maybe 10% instead of 5%.

**Comment by mhchia** [comment]
@carver No problem! Thanks a lot for the answer 🙏

**Comment by carver** [comment]
eth-keys v0.2.4 is released, so I expect `py36-rpc-blockchain` to be fixed now :+1:

File: eth2/_utils/bls_bindings/chia_network/api.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    return domain.to_bytes(8, 'big')  # bytes8
+
+
+def _privkey_int_to_bytes(privkey: int) -> bytes:
+    return privkey.to_bytes(bls_chia.PrivateKey.PRIVATE_KEY_SIZE, "big")
+
+
+def _hash_to_be_signed(message_hash: Hash32, domain: int) -> bytes:
+    domain_in_bytes = _domain_to_bytes(domain)
+    return message_hash + domain_in_bytes
```
Comment (ChihChengLiang) [inline eth2/_utils/bls_bindings/chia_network/api.py]: `_domain_to_bytes` is used only once, what do you think just do:
```
return message_hash + domain.to_bytes(8, 'big')
```

File: eth2/_utils/bls_bindings/chia_network/api.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def sign(message_hash: Hash32,
+         privkey: int,
+         domain: int) -> BLSSignature:
+    privkey_chia = bls_chia.PrivateKey.from_bytes(_privkey_int_to_bytes(privkey))
+    hash_to_be_signed = _hash_to_be_signed(message_hash, domain)
+    sig_chia = privkey_chia.sign_insecure(hash_to_be_signed)
```
Comment (ChihChengLiang) [inline eth2/_utils/bls_bindings/chia_network/api.py]: What do you think renaming `_hash_to_be_signed` to `combine_domain`?
So this line can be rewrited to:
```
sig_chia = privkey_chia.sign_insecure(combine_domain(message_hash, domain))
```

File: eth2/_utils/bls_bindings/chia_network/api.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    pubkey_chia = bls_chia.PublicKey.from_bytes(pubkey)
+    signature_chia = bls_chia.Signature.from_bytes(signature)
+    signature_chia.set_aggregation_info(
+        bls_chia.AggregationInfo.from_msg(pubkey_chia, hash_to_be_signed)
+    )
+    return cast(bool, signature_chia.verify())
+
+
+def aggregate_signatures(signatures: Sequence[BLSSignature]) -> BLSSignature:
+    signatures_chia = [
```
Comment (ChihChengLiang) [inline eth2/_utils/bls_bindings/chia_network/api.py]: Do tuple here? Looks like the test is happy with tuple.

File: eth2/_utils/bls_bindings/chia_network/api.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        bls_chia.InsecureSignature.from_bytes(signature)
+        for signature in signatures
+    ]
+    aggregated_signature = bls_chia.InsecureSignature.aggregate(signatures_chia)
+    aggregated_signature_bytes = aggregated_signature.serialize()
+    return cast(BLSSignature, aggregated_signature_bytes)
+
+
+def aggregate_pubkeys(pubkeys: Sequence[BLSPubkey]) -> BLSPubkey:
+    pubkeys_chia = list(map(bls_chia.PublicKey.from_bytes, pubkeys))
```
Comment (ChihChengLiang) [inline eth2/_utils/bls_bindings/chia_network/api.py]: Maybe use the same style in `aggregate_signatures` and `aggregate_pubkeys`?
Either all use `[bls_chia.PublicKey.from_bytes(pubkey) for pubkey in pubkeys]` or all use `tuple(map(bls_chia.InsecureSignature.from_bytes, signatures))`

**Pull request merged**